toolchain/check/handle_impl.cpp

```diff
@@ -353,13 +353,14 @@ static auto BuildImplDecl(Context& context, Parse::AnyImplDeclId node_id,
       context.node_stack().PopExprWithNodeId();
   auto [self_type_node, self_inst_id] =
       context.node_stack().PopWithNodeId<Parse::NodeCategory::ImplAs>();
-  auto self_type_id = context.types().GetTypeIdForTypeInstId(self_inst_id);
   // Pop the `impl` introducer and any `forall` parameters as a "name".
   auto name = PopImplIntroducerAndParamsAsNameComponent(context, node_id);
   auto decl_block_id = context.inst_block_stack().Pop();
 
-  // Convert the constraint expression to a type.
-  auto [constraint_inst_id, constraint_type_id] =
+  // Convert the self and constraint expressions to types.
+  auto [self_type_inst_id, self_type_id] =
+      ExprAsType(context, node_id, self_inst_id);
```

Comment:
The self expression was already converted to a type by the relevant `ImplAs` node handler; we shouldn't be doing this again. Maybe those handlers should push a `TypeInstId` instead of a plain `InstId` on the node stack?